### ISSUE TITLE
Edge routing collapse on node move

### ISSUE THREAD

**Issue by lafrank** [Microsoft/automatic-graph-layout#144]
When two nodes have multiple edges in between them, they are drawn as separate lines when performing a layout calculation. That is perfect so far.
However, when I move a node away, these edges collapse visually into a single line in between the two nodes.
How can I prevent this to happen, or how can I trigger rerouting of edges - without recalculating layout -the same way it happens with layout calculation.

**Comment by levnach** [comment]
Try holding a Shift key when you drag a node.

**Comment by lafrank** [comment]
Wow, that's it ! Cool, thank you.

**Comment by lafrank** [comment]
Just one more question. How can I achieve rerouting to happen from code ?

**Comment by lafrank** [comment]
Reopening this case as I can't find the way to achieve this functionality from code. 
So my question is : how is it possible to trigger re-routing only when a node drag ended, like if the node was ragged while holding the shift key ?

**Comment by levnach** [comment]
Please have a look at function GeometryGraphEditor. RunSplineRouterAndPutLabels(). It does the job of rerouting the edges.

**Comment by lafrank** [comment]
Thank you, that helps. 

However, it is not easy - if not impossible - to get this done from user code without modifying MSAGL, library which is not very nice. Most of the required members are not public.

To start with, I can't even catch the OnDragEnd event as it is consumed by GraphViewerGDI without publishing it. So first I need to make this event available for subscription by consumers.

It is also unclear for me, why the GraphViewGDI class is not explicitly implementing the IViewer interface, but rather casts itself as IView in certain calls :

![image](https://user-images.githubusercontent.com/15701860/31318155-e37ede9a-ac4d-11e7-9fca-79dd3e10b2d9.png)

![image](https://user-images.githubusercontent.com/15701860/31318150-d4ad1008-ac4d-11e7-9f43-5f0898a6297d.png)

I highly appreciate your help and admire the mathematics behind - read some of your publications in this topic -  but the code itself is very confusing and hard to follow.

**Comment by levnach** [comment]
I am afraid you need to modify the code of MSAGL to reach this effect. From my experience, the viewers that edit entities are extremely complex and confusing.

**Comment by lafrank** [comment]
Thanks Lev, I will manage it this way coping with complexity...

One more question, if you still had some patience...
Is there a way to influence the placement distance of a node from another node, or at least modify the node size ?
The below illustrates a part of the graph my application currently produces:
![image](https://user-images.githubusercontent.com/15701860/31353767-18a5cce6-ad34-11e7-81cc-bfbbec6c208b.png)
The bigger picture is:
![image](https://user-images.githubusercontent.com/15701860/31353892-976664c8-ad34-11e7-8827-5f2c858a88e3.png)


As you can see,, some anomalies stem fromt the fact that the gray - label only - nodes were placed too far from its neighbor node by MDS algorithm with current settings.

What I would like to achive instead is something like below:

![image](https://user-images.githubusercontent.com/15701860/31353981-e400a7ee-ad34-11e7-81d7-7fb820ceec19.png)

How is it possible ?

Many thanks in advance.

**Comment by levnach** [comment]
The simplest way is to change the font size of the "center nodes", something like that  
graph.FindNode("47").Label.FontSize *= 3;
If it does not look good  you can get control over the node size and rendering as in NodesWithImages sample.

**Comment by lafrank** [comment]
No success with this unfortunately.  Although I don't understand MDS in depth, the layout generated doesn't make sense in my view :

![image](https://user-images.githubusercontent.com/15701860/31371830-588fc5f0-ad93-11e7-87bb-0cad1673c167.png)

**Comment by levnach** [comment]
MDS attempts to position the nodes that have similar set neighbors close by. The edge directions are ignored. If your graph represents some process and has directed edges then Sugiyama scheme is a better option.
But the picture above does not look right, the labels on large nodes are too small. It could be a bug.

**Comment by lafrank** [comment]
What the graph represents is a computer network with the blue nodes being routers. They have their labels cleared for posting due to company privacy considerations, sorry for causing confusion with this.


The problem is that the grey, text-only nodes - visible on the top-left corner - have only one node connected to them, and still are placed very far from their single connected node.

Just as you explained MDS, my goal would be the same, having single edge interconnected nodes as close to each other as possible. And wondering why this is not happening ? 

If you look at the below once again, you can see exactly what my issue is :
![image](https://user-images.githubusercontent.com/15701860/31397185-a2c527e4-ade5-11e7-82a3-4b06fe756e8b.png)


Edge diretions are not important to me at this stage.

**Comment by levnach** [comment]
Can you try another layout, the one that is used in FastIncrementalLayoutWithGdi?
It might work better for you.

**Comment by lafrank** [comment]
Sure, I will soon, and send an update...

**Comment by lafrank** [comment]
I was experimenting a lot with clusters, but was not able to achive a good-looking layout un fortunately.
Are clusters only working with IncrementalLayout or also with MDS ?
And one more question : what is the easiest way to update a node position from code, like it was moved away by dragging it ? I am asking because the Pos property is readn only.

**Comment by lafrank** [comment]
I also tried  

`gViewer1.LayoutEditor.geomGraphEditor.RunSplineRouterAndPutLabels();`

but this does not have any visible effect.

**Comment by lafrank** [comment]
I am closing this issue as most of my concerns are addressed in pull request #150

Thanks @levnach once again for your help so far.
